flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -661,18 +664,31 @@ private void configureCheckpointing() {
 				"exactly-once or at-least-once.");
 		}
 
-		//  --- configure the master-side checkpoint hooks ---
+		//  --- configure the master-side checkpoint hooks and check the types of operators---
 
 		final ArrayList<MasterTriggerRestoreHook.Factory> hooks = new ArrayList<>();
 
 		for (StreamNode node : streamGraph.getStreamNodes()) {
-			if (node.getOperatorFactory() instanceof UdfStreamOperatorFactory) {
-				Function f = ((UdfStreamOperatorFactory) node.getOperatorFactory()).getUserFunction();
+			StreamOperatorFactory operatorFactory = node.getOperatorFactory();
+			if (operatorFactory instanceof UdfStreamOperatorFactory) {
+				Function f = ((UdfStreamOperatorFactory) operatorFactory).getUserFunction();
 
 				if (f instanceof WithMasterCheckpointHook) {
 					hooks.add(new FunctionMasterCheckpointHookFactory((WithMasterCheckpointHook<?>) f));
 				}
 			}
+
+			if (cfg.isCheckpointingEnabled() && operatorFactory != null) {
```

Comment:
I'd move this check (for all nodes) into a separate function and invoke the new function earlier (in `StreamingJobGraphGenerator.createJobGraph()` call or even before `StreamingJobGraphGenerator` is used).

`StreamGraph.getJobGraph()` already has a validation check before creating a `JobGraph`. Imo, it's a better place to carry such validation for now (excluding the case for `InputSelectable` and network credit-based flow control).

Reply:
The following validation that you are talking about allows to enable the forced checkpointing for the iterative streaming, which means that no exception is thrown in that case. In this PR, we need to add the check to cover this case, otherwise `IterateITCase#testWithCheckPointing` will fail.
```
if (isIterative() && checkpointConfig.isCheckpointingEnabled() && !checkpointConfig.isForceCheckpointing()) 
```

Reply:
I agree with @1u0. This method is definitely already too big if we want to add some new functionality here, it must be decomposed into smaller pieces. But I also agree that this check would be better to perform earlier. Maybe in some `private void  StreamingJobGraphGenerator#validateJobGraph()` method (together with the present validation that happens in `StreamGraph#getJobGraph`). 

With this `validateJobGraph()` either we could call it pre-emptively as a first thing in `createJobGraph()` or expect a user to call it manually before `createJobGraph()`.